coremltools/converters/mil/mil/passes/defs/optimize_normalization.py

```diff
@@ -823,6 +836,190 @@ def _try_match_and_transform_pattern_4(self, reduce_op: Operation, block: Block)
+        ):
+            return False
+        if not (sub_op1.x == root_var and sub_op1.y == reduce_op.outputs[0]):
+            return False
+        ops_to_remove.append(sub_op1)
+
+        # check mul op (equivalent to a square op)
+        square_op = self._try_get_child_op_type(sub_op1, "mul")
+        if square_op is None or not self._check_child_op_types(
+            square_op, child_op_types=["reduce_mean"]
+        ):
+            return False
+        if square_op.x != square_op.y:
+            return False
+        ops_to_remove.append(square_op)
+
+        # check second reduce mean
+        reduce_op2 = self._try_get_child_op_type(square_op, "reduce_mean")
+        if not self._check_reduce_op(reduce_op2) or not self._check_child_op_types(
+            reduce_op2, child_op_types=["add"]
+        ):
+            return False
+        if len(reduce_op2.axes.val) != 1 or reduce_op2.axes.val != [1] or not reduce_op2.keep_dims.val:
+            return False
+        ops_to_remove.append(reduce_op2)
+
+        # check add op (with epsilon)
+        add_op1 = self._try_get_child_op_type(reduce_op2, "add")
+        if add_op1 is None or not self._check_child_op_types(
+            add_op1, child_op_types=["rsqrt"]
+        ):
+            return False
+        epsilon_var = add_op1.y if add_op1.x == reduce_op2.outputs[0] else add_op1.x
+        if epsilon_var.val is None or len(epsilon_var.val.shape) != 0:
+            return False  # must be scalar
+        ops_to_remove.append(add_op1)
+
+        # check rsqrt
+        rsqrt_op = self._try_get_child_op_type(add_op1, "rsqrt")
+        if rsqrt_op is None or not self._check_child_op_types(
+            rsqrt_op, child_op_types=["mul"]
+        ):
+            return False
+        ops_to_remove.append(rsqrt_op)
+
+        # Last op in pattern if there is no elementwise affine.
+        mul_op = self._try_get_child_op_type(rsqrt_op, "mul")
+        if mul_op is None:
+            return False
+        if mul_op.y != sub_op1.outputs[0] and mul_op.x != sub_op1.outputs[0]:
+            return False
+        ops_to_remove.append(mul_op)
+
+        # Default values if no gamma or beta ops.
+        end_op = mul_op
+        gamma_var = None
+        beta_var = None
+
+        add_beta_op = self._try_get_child_op_type(mul_op, "add")
+        mul_gamma_op = self._try_get_child_op_type(add_beta_op, "mul")
+
+        has_beta_and_gamma = add_beta_op is not None and mul_gamma_op is not None
+
+        if has_beta_and_gamma and not self._check_child_op_types(
+            mul_op, child_op_types=["add"]
+        ):
+            # It would be possible to fuse this pattern as:
+            # layer_norm(x, gamma=None, beta=None) -> add(beta) -> mul(gamma) -> ...
+            #                                      |-> other mul_op child ops
+            # For simplicity don't handle this edge case.
+            return False
+
+        if has_beta_and_gamma and self._check_child_op_types(
```

Comment:
Same here,
if `has_beta_and_gamma is True`, that already guarantees  `self._check_child_op_types(add_beta_op, child_op_types=["mul"] is True`?

Reply:
Similarly, if `add_beta_op` were an input to some downstream op then fusing would break it. I'll refactor this to be a little clearer and add a test case for it.

Reply:
Great! Thanks!